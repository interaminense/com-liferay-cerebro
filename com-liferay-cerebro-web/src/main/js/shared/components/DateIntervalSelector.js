import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';
import {mapValues, noop} from 'lodash';

import Button from './Button';
import DatePicker from './date-picker';
import Icon from './Icon';
import Overlay from './Overlay';
import RadioGroup from './RadioGroup';
import SearchableSelect from './SearchableSelect';

export const RANGE = {
	CUSTOM: 'CUSTOM',
	LAST_MONTH: 'LAST_MONTH',
	LAST_THREE_MONTHS: 'LAST_THREE_MONTHS',
	LAST_WEEK: 'LAST_WEEK',
	LAST_YEAR: 'LAST_YEAR'
};

export const INTERVAL = {
	DAILY: 'DAILY',
	MONTHLY: 'MONTHLY',
	QUARTERLY: 'QUARTERLY',
	WEEKLY: 'WEEKLY',
	YEARLY: 'YEARLY'
};

const RANGE_LANG_MAP = {
	[RANGE.CUSTOM]: Liferay.Language.get('custom'),
	[RANGE.LAST_MONTH]: Liferay.Language.get('last-month'),
	[RANGE.LAST_THREE_MONTHS]: Liferay.Language.get('last-three-months'),
	[RANGE.LAST_WEEK]: Liferay.Language.get('last-week'),
	[RANGE.LAST_YEAR]: Liferay.Language.get('last-year')
};

const INTERVAL_LANG_MAP = {
	[INTERVAL.DAILY]: Liferay.Language.get('daily'),
	[INTERVAL.MONTHLY]: Liferay.Language.get('monthly'),
	[INTERVAL.QUARTERLY]: Liferay.Language.get('quarterly'),
	[INTERVAL.WEEKLY]: Liferay.Language.get('weekly'),
	[INTERVAL.YEARLY]: Liferay.Language.get('yearly')
};

const INTERVAL_OPTIONS = [
	INTERVAL.DAILY,
	INTERVAL.WEEKLY,
	INTERVAL.MONTHLY,
	INTERVAL.QUARTERLY,
	INTERVAL.YEARLY
];

const RANGE_OPTIONS = [
	RANGE.LAST_WEEK,
	RANGE.LAST_MONTH,
	RANGE.LAST_THREE_MONTHS,
	RANGE.LAST_YEAR,
	RANGE.CUSTOM
].map(value => ({
	name: RANGE_LANG_MAP[value],
	value
}));

class DateIntervalSelector extends Component {
	getIntervalLabel() {
		const {interval} = this.props;

		return INTERVAL_LANG_MAP[interval];
	}

	getRangeLabel() {
		const {buttonDateFormat, range, rangeType} = this.props;

		if (rangeType === RANGE.CUSTOM) {
			const {start, end} = mapValues(
				range,
				value => (value ? value.format(buttonDateFormat) : '****')
			);

			return `${start} - ${end}`;
		}

		return RANGE_LANG_MAP[rangeType];
	}

	@autobind
	handleRangeTypeSelect(item) {
		this.props.onRangeTypeChange(item.value);
	}

	@autobind
	handleToggleActive() {
		this.state.active = !this.state.active;
	}

	render() {
		const {
			props: {
				interval,
				onIntervalChange,
				onRangeChange,
				range,
				rangeType
			},
			state: {active}
		} = this;

		const classes = 'date-interval-selector-root';

		return (
			<Overlay
				active={active}
				containerClass={classes}
				onOutsideClick={this.handleToggleActive}
			>
				<Button
					display="unstyled"
					elementClasses={classes}
					onClick={this.handleToggleActive}
				>
					<div class="label-container">
						{this.getRangeLabel()}

						<span class="button-label-separator">{'|'}</span>

						<span class="interval-label">
							{this.getIntervalLabel()}
						</span>
					</div>

					<div class="icon-container">
						<Icon symbol="calendar" />
					</div>
				</Button>

				<div class="date-interval-selector-content">
					<div class="interval-container">
						<div class="form-group">
							<label>{Liferay.Language.get('date-range')}</label>

							<SearchableSelect
								buttonProps={{
									size: 'sm'
								}}
								items={RANGE_OPTIONS}
								onSelect={this.handleRangeTypeSelect}
								selectedItem={{
									name: RANGE_LANG_MAP[rangeType],
									value: rangeType
								}}
								showSearch={false}
							/>
						</div>

						<div class="form-group">
							<label>
								{Liferay.Language.get('time-interval')}
							</label>

							<RadioGroup
								checked={interval}
								onChange={onIntervalChange}
							>
								{INTERVAL_OPTIONS.map(value => (
									<RadioGroup.Option
										key={value}
										label={INTERVAL_LANG_MAP[value]}
										value={value}
									/>
								))}
							</RadioGroup>
						</div>
					</div>

					<div class="date-picker-container">
						<DatePicker
							date={range}
							disabled={rangeType !== RANGE.CUSTOM}
							onSelect={onRangeChange}
						/>
					</div>
				</div>
			</Overlay>
		);
	}
}

DateIntervalSelector.STATE = {
	active: Config.bool().value(false)
};

DateIntervalSelector.PROPS = {
	buttonDateFormat: Config.string().value('MMM D, YYYY'),
	interval: Config.oneOf(Object.values(INTERVAL)).value(INTERVAL.LAST_MONTH),
	onIntervalChange: Config.func().value(noop),
	onRangeChange: Config.func().value(noop),
	onRangeTypeChange: Config.func().value(noop),
	range: Config.object(),
	rangeType: Config.oneOf(Object.values(RANGE)).value(RANGE.DAILY)
};

export default DateIntervalSelector;