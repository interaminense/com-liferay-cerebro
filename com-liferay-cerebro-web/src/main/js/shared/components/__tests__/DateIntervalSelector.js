import moment from 'moment';

import DateIntervalSelector, {RANGE, INTERVAL} from '../DateIntervalSelector';
import {renderWithStore} from 'test/mock-store';
import renderWithOverlay from 'test/overlay';

describe('DateIntervalSelector', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render', () => {
		component = renderWithStore(DateIntervalSelector);

		expect(component).toMatchSnapshot();
	});

	it('should render content in overlay', () => {
		component = renderWithOverlay(DateIntervalSelector);

		component.refs.child.handleToggleActive();

		jest.runAllTimers();

		expect(component).toMatchSnapshot();
	});

	it('should render button label with custom range', () => {
		component = renderWithStore(DateIntervalSelector, {
			range: {
				end: moment(1000000000),
				start: moment(0)
			},
			rangeType: RANGE.CUSTOM
		});

		expect(component).toMatchSnapshot();
	});

	it('should render button label with different interval', () => {
		component = renderWithStore(DateIntervalSelector, {
			interval: INTERVAL.YEARLY
		});

		expect(component).toMatchSnapshot();
	});

	it('should call onRangeTypeChange', () => {
		const onRangeTypeChange = jest.fn();

		component = renderWithStore(DateIntervalSelector, {
			onRangeTypeChange
		});

		expect(onRangeTypeChange).not.toHaveBeenCalled();

		component.handleRangeTypeSelect({value: 0});

		expect(onRangeTypeChange).toHaveBeenCalledWith(0);
	});
});