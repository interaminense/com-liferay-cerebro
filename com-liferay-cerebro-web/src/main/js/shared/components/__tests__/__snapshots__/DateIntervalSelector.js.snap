// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateIntervalSelector should render 1`] = `
<button id="overlay1" class="button-root btn btn-unstyled date-interval-selector-root " type="button">
  <div class="label-container">
    <span class="button-label-separator">|</span>
    <span class="interval-label"></span>
  </div>
  <div class="icon-container">
    <svg class="icon-root lexicon-icon lexicon-icon-calendar">
      <use xlink:href="/images/lexicon/icons.svg#calendar"></use>
    </svg>
  </div>
</button>
`;

exports[`DateIntervalSelector should render button label with custom range 1`] = `
<button id="overlay4" class="button-root btn btn-unstyled date-interval-selector-root " type="button">
  <div class="label-container">Dec 31, 1969 - Jan 12, 1970
    <span class="button-label-separator">|</span>
    <span class="interval-label"></span>
  </div>
  <div class="icon-container">
    <svg class="icon-root lexicon-icon lexicon-icon-calendar">
      <use xlink:href="/images/lexicon/icons.svg#calendar"></use>
    </svg>
  </div>
</button>
`;

exports[`DateIntervalSelector should render button label with different interval 1`] = `
<button id="overlay5" class="button-root btn btn-unstyled date-interval-selector-root " type="button">
  <div class="label-container">
    <span class="button-label-separator">|</span>
    <span class="interval-label">yearly</span>
  </div>
  <div class="icon-container">
    <svg class="icon-root lexicon-icon lexicon-icon-calendar">
      <use xlink:href="/images/lexicon/icons.svg#calendar"></use>
    </svg>
  </div>
</button>
`;

exports[`DateIntervalSelector should render content in overlay 1`] = `
<span>
  <div class="overlay-renderer-root">
    <div data-overlayid="2" id="overlay2Container" class="card overlay-container-root date-interval-selector-root" style="top: 0px; left: 0px;">
      <div class="date-interval-selector-content">
        <div class="interval-container">
          <div class="form-group">
            <label>date-range</label>
            <div class="dropdown dropdown-root searchable-select-root" id="overlay3">
              <button class="button-root btn btn-secondary btn-sm dropdown-toggle" type="button">
                <span class="text-truncate"></span>
                <span>
                  <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
                    <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
                  </svg>
                </span>
              </button>
            </div>
          </div>
          <div class="form-group">
            <label>time-interval</label>
            <div class="form-check">
              <div class="custom-control custom-radio">
                <label>
                  <input class="custom-control-input" type="radio">
                  <span class="custom-control-indicator"></span>
                  <span class="custom-control-description">daily</span>
                </label>
              </div>
              <div class="custom-control custom-radio">
                <label>
                  <input class="custom-control-input" type="radio">
                  <span class="custom-control-indicator"></span>
                  <span class="custom-control-description">weekly</span>
                </label>
              </div>
              <div class="custom-control custom-radio">
                <label>
                  <input class="custom-control-input" type="radio">
                  <span class="custom-control-indicator"></span>
                  <span class="custom-control-description">monthly</span>
                </label>
              </div>
              <div class="custom-control custom-radio">
                <label>
                  <input class="custom-control-input" type="radio">
                  <span class="custom-control-indicator"></span>
                  <span class="custom-control-description">quarterly</span>
                </label>
              </div>
              <div class="custom-control custom-radio">
                <label>
                  <input class="custom-control-input" type="radio">
                  <span class="custom-control-indicator"></span>
                  <span class="custom-control-description">yearly</span>
                </label>
              </div>
            </div>
          </div>
        </div>
        <div class="date-picker-container">
          <div aria-disabled="" class="date-picker-root disabled">
            <div class="controls">
              <div class="month-toggle-wrapper">
                <button class="button-root btn btn-secondary btn-sm" type="button">November 2017
                  <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
                    <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
                  </svg>
                </button>
              </div>
              <button class="button-root btn btn-secondary btn-sm btn-monospaced" type="button">
                <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
                  <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
                </svg>
              </button>
              <button disabled="" class="button-root btn btn-secondary btn-sm btn-monospaced current-day-btn" type="button">â€¢</button>
              <button class="button-root btn btn-secondary btn-sm btn-monospaced" type="button">
                <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
                  <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
                </svg>
              </button>
            </div>
            <div class="picker-body">
              <table class="calendar-root">
                <thead>
                  <tr>
                    <th>Sun</th>
                    <th>Mon</th>
                    <th>Tue</th>
                    <th>Wed</th>
                    <th>Thu</th>
                    <th>Fri</th>
                    <th>Sat</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root outside-month" type="button">29</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root outside-month" type="button">30</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root outside-month" type="button">31</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">1</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">2</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">3</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">4</button>
                      </div>
                    </td>
                  </tr>
                  <tr>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">5</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">6</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">7</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">8</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">9</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">10</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">11</button>
                      </div>
                    </td>
                  </tr>
                  <tr>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">12</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">13</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">14</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">15</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">16</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">17</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">18</button>
                      </div>
                    </td>
                  </tr>
                  <tr>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">19</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">20</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">21</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">22</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">23</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">24</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">25</button>
                      </div>
                    </td>
                  </tr>
                  <tr>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">26</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">27</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">28</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">29</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root" type="button">30</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root outside-month" type="button">1</button>
                      </div>
                    </td>
                    <td class="">
                      <div class="day-container">
                        <button class="button-root btn btn-unstyled day-root outside-month" type="button">2</button>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button id="overlay2" class="button-root btn btn-unstyled active date-interval-selector-root " type="button">
    <div class="label-container">
      <span class="button-label-separator">|</span>
      <span class="interval-label"></span>
    </div>
    <div class="icon-container">
      <svg class="icon-root lexicon-icon lexicon-icon-calendar">
        <use xlink:href="/images/lexicon/icons.svg#calendar"></use>
      </svg>
    </div>
  </button>
</span>
`;
